docs/source/config.rst
Configuration Options
#####################

Enable openNDS
**************

Set to 0 to disable opennds

``option enabled 1``

Use deprecated generic configuration file
******************************************

Use of this setting is no longer supported and if present is ignored.

``option config '/etc/opennds/opennds.conf'``

Enable debug output (0-3)
*************************

Default: 1

Level0
------
Silent (only initial startup, LOG_ERR and LOG_EMERG messages will be seen, otherwise there will be no logging.)

Level 1
-------
LOG_ERR, LOG_EMERG, LOG_WARNING and LOG_NOTICE (this is the default level).

Level 2
-------
debuglevel 1  + LOG_INFO

Level 3
-------
debuglevel 2 + LOG_DEBUG

``option debuglevel '1'``

Firewall Restart hook
*********************

Specific to OpenWrt only, Firewall 4 (FW4) informs openNDS when it is restarting.

If enabled (Set to 1), openNDS reinserts any nftables it may need in the FW4 ruleset.

Default: 1

``option fwhook_enabled '1'``

Ndsctl Socket Access
********************

Set the socket name to use for ndsctl socket access, relative to the tmpfs mountpoint.

Any directory/folder specified must exist.
Default: ndsctl.sock (Do not add a leading "/")

Full default socket path would be /tmp/ndsctl.sock in OpenWrt
In the following example, the socket path would be /tmp/sockets/ndsctl.sock

Example:

``option ndsctlsocket 'sockets/ndsctl.sock'``

Local Log Mountpoint
********************

Default: router's volatile tmpfs storage eg on OpenWrt '/tmp'

Local logging can be directed to any storage accessible to the router eg USB drive, SSD etc

**WARNING** - you should not use the router's built in flash storage as this would cause
excessive wear and in a live system will result quite quickly in flash failure making the router useless.

Example:

``option log_mountpoint '/logs'``

Maximum number of Local Log Entries
***********************************

Set the maximum number of local log entries to be kept.
Default 100

Minimum value 0 (no limit)

Maximum value - limited only by free storage space on the logging mountpoint

If set to '0' there is no limit

This is the maximum number of local log entries allowed before log rotation begins

Both ThemeSpec and BinAuth log locally if they are enabled

**WARNING** - local logging is by default written to the tmpfs volatile storage

If this option were to be set too high the router could run out of tmpfs storage and/or free RAM

Non-volatile storage, such as a USB storage device may be defined using the log_mountpoint option

Example:

``option max_log_entries '1000'``

Login Option
************

Default: 1

Integer value sent to PreAuth script as login mode

openNDS comes preconfigured for three basic modes of operation

Mode 0
------
If FAS is not enabled, then this functions as mode 1

If FAS is configured, FAS is used.

Mode 1
------
Default Dynamic Click to Continue

The pre-installed dynamic login page is enabled by setting option login_option_enabled = '1'

It generates a Click to Continue page followed by an info/advertising page.

User clicks on “Continue” are recorded in the log file /[tmpfs_dir]/ndslog/ndslog.log

Mode 2
------
Username/Emailaddress Dynamic Login

The pre-installed dynamic login page is enabled by setting option login_option_enabled = '2'

It generates a login page asking for username and email address followed by an info/advertising page.

User logins are recorded in the log file /[tmpfs_dir]/ndslog/ndslog.log

Mode 3
------
Use Theme defined in ThemeSpec path (option themespec_path)

`option login_option_enabled '1'`

Allow Preemptive Authentication
*******************************

Default: 1 - Enabled

Disable by setting to 0

This allows the ndsctl utility to preemptively authorise **connected** clients that have not entered the preauthenticated state.

This is useful for example with IoT devices that do not have CPD (captive portal detection)

or for a FAS to manage inter-captive-portal roaming by making use of a centralised database of client validations.

Example:

``option allow_preemptive_authentication '0'``

ThemeSpec Path
**************

Default: None

Required when when login_option_enabled is set to '3'

Note: /usr/lib/opennds/theme_click-to-continue-basic.sh is used for login_option_enabled '1'

and:  /usr/lib/opennds/theme_user-email-login-basic.sh is used for login_option_enabled '2'

Sets the ThemeSpec file path to be used when login_option_enabled '3'

The ThemeSpec script makes use of lists of custom parameters, custom variables, custom image urls and custom files and is used to generate the dynamic splash page sequence.

The ThemeSpec file will normally reside in /usr/lib/opennds/ but can be anywhere accessible to openNDS.

The file must be flagged as executable and have the correct shebang for the default shell.

``option themespec_path '/usr/lib/opennds/<filename>'``

DHCP Leases File
****************

Default: Try /tmp/dhcp.leases, /var/lib/misc/dnsmasq.lease or /var/db/dnsmasq.leases

The file containing the list of active DHCP leases.

Example:

``option dhcp_leases_file '/tmp/dhcp.leases.special'``

Define Custom Parameters
************************

Custom parameters are sent as fixed values to FAS

Default None

Custom Parameters listed in the form of param_name=param_value

param_name and param_value must be urlencoded if containing white space or single quotes

eg replace spaces with %20 - replace single quotes with %27

Parameters should be configured one per line to prevent possible parsing errors.

eg:

``list fas_custom_parameters_list '<param_name1=param_value1>'``

``list fas_custom_parameters_list '<param_name2=param_value2>'``

etc.

Configuration for custom parameters in the installed ThemeSpec Files
--------------------------------------------------------------------

The installed ThemeSpec files are:

theme_click-to-continue-custom-placeholders

and

theme_user-email-login-custom-placeholders

``list fas_custom_parameters_list 'logo_message=openNDS:%20Perfect%20on%20OpenWrt!'``

``list fas_custom_parameters_list 'banner1_message=BlueWave%20-%20Wireless%20Network%20Specialists'``

``list fas_custom_parameters_list 'banner2_message=HMS%20Pickle'``

``list fas_custom_parameters_list 'banner3_message=SeaWolf%20Cruiser%20Racer'``

Define Custom Variables
***********************

Custom Variables are used by FAS to dynamically collect information from clients

Default None

Custom Variables are listed in the form of var_name=var_type

"var_name" and "var_type" must be urlencoded if containing white space or single quotes

eg replace spaces with %20 - replace single quotes with %27

Variables should be configured one per line to prevent possible parsing errors.

eg:

``list fas_custom_variables_list '<var_name1=var_type1>'``

``list fas_custom_variables_list '<var_name2=var_type2>'``

etc.

FAS Generic Variables
---------------------
A custom FAS or ThemeSpec must be written to make use of FAS Generic Variables

eg:

``list fas_custom_variables_list 'membership_number=number'``

``list fas_custom_variables_list 'access_code=password'``

ThemeSpec Dynamically generated Form Fields
-------------------------------------------

ThemeSpec scripts can dynamically generate Form Field html and inject into the dynamic splash page sequence.

This is achieved using a SINGLE line containing the keyword "input", in the form: fieldname:field-description:fieldtype

Numerous fields can be defined in this single "input=" line, separated by a semicolon (;).

Configuration for custom variables in the installed ThemeSpec Files
-------------------------------------------------------------------

theme_click-to-continue-custom-placeholders

and

theme_user-email-login-custom-placeholders

This example inserts Phone Number and Home Post Code fields:

``list fas_custom_variables_list 'input=phone:Phone%20Number:text;postcode:Home%20Post%20Code:text'``

Define Custom Images
********************

Custom Images are served by a local FAS where required in dynamic portal pages

Default None

Custom images will be copied from the URL to the openNDS router

Custom Images are listed in the form of image_name_type=image_url

image_name and image_url must be urlencoded if containing white space or single quotes

The image url must begin with http:// https:// or file://

Images should be configured one per line to prevent possible parsing errors.

``list fas_custom_images_list '<image_name1_[type]=image_url1>'``

``list fas_custom_images_list '<image_name2_[type]=image_url2>'``

etc.

"type" can be any recognised image file extension eg jpg, png, ico, etc.

Configuration for custom images in the installed ThemeSpec Files
----------------------------------------------------------------

theme_click-to-continue-custom-placeholders

and

theme_user-email-login-custom-placeholders

``list fas_custom_images_list 'logo_png=https://openwrt.org/_media/logo.png'``

``list fas_custom_images_list 'banner1_jpg=https://raw.githubusercontent.com/openNDS/openNDS/v9.0.0/resources/bannerbw.jpg'``

``list fas_custom_images_list 'banner2_jpg=https://raw.githubusercontent.com/openNDS/openNDS/v9.0.0/resources/bannerpickle.jpg'``

``list fas_custom_images_list 'banner3_jpg=https://raw.githubusercontent.com/openNDS/openNDS/v9.0.0/resources/bannerseawolf.jpg'``

Define Custom Files
*******************

Custom Files are served by a local FAS where required in dynamic portal pages

Default None

Custom files will be copied from the URL to the openNDS router

Images should be configured one per line to prevent possible parsing errors.

Custom files are listed in the form of file_name_type=file_url

file_name and file_url must be urlencoded if containing white space or single quotes

The file url must begin with http:// https:// or file://

``list fas_custom_files_list '<file_name1_[type]=file_url1>'``

``list fas_custom_files_list '<file_name2_[type]=file_url2>'``

"type" can be any recognised file extension that can be used to display web content eg txt, htm etc.

URLs using the file:// protocol must point to a valid mountpoint accessible to openNDS, for example a usb storage device.

Configuration for custom files in the installed ThemeSpec Files
----------------------------------------------------------------

theme_click-to-continue-custom-placeholders

and

theme_user-email-login-custom-placeholders

``list fas_custom_files_list 'advert1_htm=https://raw.githubusercontent.com/openNDS/openNDS/v9.0.0/resources/bannerpickle.htm'``


Set refresh interval for downloads
**********************************

Set refresh interval for downloaded remote files (in minutes)

Default 0

A setting of 0 (zero) means refresh is disabled.

This is useful for providing automated refreshing of informational or advertising content. Should the remote resources become unavailable, current versions will continue to be used.

Example, set to twelve hours (720 minutes):

``option remotes_refresh_interval '720'``

Use outdated libmicrohttpd (MHD)
********************************

Default 0 (Disabled)

**Warning**: enabling this *may* cause instability or in the worst case total failure - it would be better to upgrade MHD.

**Use at your own risk.**

Older versions of MHD use an older version of the MHD API and may not run correctly or fail.

Older versions of MHD convert & and + characters to spaces when present in form data. This can make a PreAuth or BinAuth impossible to use for a client if form data contains either of these characters eg. in username or password.

*There may well be other issues with older versions.*

MHD versions earlier than 0.9.71 are detected.

If this option is set to 0 (default), NDS will terminate if MHD is earlier than 0.9.71

If this option is set to 1, NDS will attempt to start and log an error.

``option use_outdated_mhd '1'``

Maximum Page Size to be served by MHD
*************************************

Default 10240 bytes

Minimum value 1024 bytes

Maximum - limited only by free RAM in the router

This sets the maximum number of bytes that will be served per page by the MHD web server.

Setting this option is useful:

	1. To reduce memory requirements on a resource constrained router
	2. To allow large pages to be served where memory usage is not a concern

Example:

``option max_page_size '4096'``

Set the MHD WebRoot
*******************

Default: /etc/opennds/htdocs

The local path where the system css file, and other static page content resides.

ie. Serve the file splash.css from this directory

Example:

``option webroot '/etc/opennds/htdocs'``

Set the GatewayInterface
************************

Default: br-lan

Use this option to set the device opennds will bind to.

The value may be an interface section in /etc/config/network or a device name such as br-lan.

The selected interface must be allocated an IPv4 address.

In OpenWrt this is normally br-lan, in generic Linux it might be wlan0

``option gatewayinterface 'br-lan'``

Set the GatewayPort
*******************

Default: 2050

openNDS's own http server (MHD) uses the gateway address as its IP address.

This option sets the port it listens to.

Example:

``option gatewayport '2080'``

Set the GatewayName
*******************

Default: openNDS

gatewayname is used as an identifier for the instance of openNDS

It is displayed on the default splash page sequence for ThemeSpec and the example php scripts.


ChangeLog
opennds (11.0.0)
  * Fix - Preemptivemac not taking config values [bluewavenet]
  * Fix - remove any trailing newlines on custom string [bluewavenet]
  * Add - ensure gatewayinterface is not a wireless interface [bluewavenet]
  * Fix - querystr buffer overflow [bluewavenet]
  * Fix - rename custombinauth_reauth_interval to check_reauth_interval [bluewavenet]
  * Fix - block potential OS Command Injection [bluewavenet]
  * Add - check_reauth_interval binauth extension [bluewavenet]
  * Fix - customdata handling [bluewavenet]
  * Fix - remove more unneeded leading spaces [bluewavenet]
  * Add - ThemeSpec theme_allow_all [bluewavenet]
  * Fix - Remove unneeded leading space in authstr and b64 encode customdata [bluewavenet]
  * Fix - inadvertant memory overwrite when customdata is null [bluewavenet]
  * Fix - typo in config file [bluewavenet]
  * Add - wrapper for calling OpenNDS libraries [bluewavenet]
  * Add - refactored theme_trusted_only [bluewavenet]
  * Fix - change fast_client_scan default to enabled [bluewavenet]
  * Add - Theme Trusted Only based on Theme Click to Continue [bluewavenet]
  * Fix - Check for correct config format on startup and abort if not [bluewavenet]
  * Fix - workaround for Android refusing to resolve local fqdn status.client [bluewavenet]
  * Fix - increase procd respawn parameters to prevent fail to restart on command [bluewavenet]
  * Fix - unnecessary memory use and possible out of bounds string access [bluewavenet]
  * Fix - some duplicated nft rules [bluewavenet]
  * Add - tcp to essentials user_to_router [bluewavenet]
  * Fix - update heartbeat always when in check_mhd [bluewavenet]
  * Add - random query string when accessing status page [bluewavenet]
  * Fix - themespec after changes to remote fas config [bluewavenet]
  * Add - Simplification of construct_querystring [bluewavenet]
  * Add - default fas_port set to 443 [bluewavenet]
  * Add - faskey update for testing purposes in fas-aes.php [bluewavenet]
  * Add - refactor checks for fasremoteip and fasremotefqdn [bluewavenet]
  * Fix - error message if mesh11sd is not running [bluewavenet]
  * Add - Error page to status.client [bluewavenet]
  * Fix - delete old nftset form config on startup [bluewavenet]
  * Add - updated defaults for fasport and login_option_enabled [bluewavenet]
  * Fix - Add MHD custom log and MHD_quiesce to prevent possible deadlock [bluewavenet]
  * Fix - Multiple changes to walled garden and blocklist sets [bluewavenet]
  * Fix - remove redundant calloc for client type [bluewavenet]
  * Fix - remove legacy ipv6_routing block [bluewavenet]
  * Add - additional ping tests for upstream online status [bluewavenet]
  * Add - additional ping tests for upstream online status [bluewavenet]
  * Fix - Free without calloc on restart if client not in client list [bluewavenet]
  * Add - ipv6 support [bluewavenet]
  * Fix - MHD version check rework [mcassaniti]
  * Add - use library function to write to syslog [bluewavenet]
  * Add - get reauth_interval from config [bluewavenet]
  * Add - list valid variables [bluewavenet]
  * Fix - set quota value to 0 (zero) if null [bluewavenet]
  * Fix - binauth_log format [bluewavenet]
  * Add - config option custombinauth and deprecate option binauth [bluewavenet]
  * Add - updated custombinauth script [bluewavenet]
  * Add - reauth_interval script [bluewavenet]
  * Fix - prevent duplication of dhcp_option_force on restart [bluewavenet]
  * Add - support binauth defined quotas for all authentication types [bluewavenet]
  * Fix - prevent status page from causing browser tabs accumilating [bluewavenet]
  * Add - client quota variables to the binauthlog entry [bluewavenet]
  * Fix - consistantly use sessiontimeout rather than session_length or session_timeout [bluewavenet]
  * Fix - don't reuse strtok buffer until we are finished with it [bluewavenet]
  * Fix - remove unused variable [bluewavenet]
  * Add - use named section in config [bluewavenet]
  * Fix - cidfile error in binauth if preemptive auth [bluewavenet]
  * Fix - overwrite of ipclient and macclient on preemptive auth [bluewavenet]
  * Fix - populate config->themespec_path when default themespecs are used [bluewavenet]
  * Fix - some edge case error outputs from libopennds [bluewavenet]
  * Add - some debug messages when freeing client node [bluewavenet]
  * Fix - memory leak in _client_free_node() [bluewavenet]
  * Fix - remove redundant client_reset() [bluewavenet]
  * Fix - pointer overwrite and memory leak setting some client parameters [bluewavenet]
  * Fix - remove redundant call to client_reset() [bluewavenet]
  * Fix - return error code from binauth script [bluewavenet]
  * Fix - support binauth quota settings [bluewavenet]
  * Fix - get exit code from bunauth script [bluewavenet]
  * Add - get_quotas_by_mac library function [bluewavenet]
  * Fix - typo in auth_file check [bluewavenet]
  * Fix - procd failure if daemon outputs syslog before procd is ready [bluewavenet]
  * Fix - potential error output if auth_file does not exist [bluewavenet]
  * Add - option fast_client_scan to inhibit scanning of mesh for client connections [bluewavenet]
  * Fix - use cidfile only if using themespec including default [bluewavenet]
  * Fix - speed up themespec html generation on low speed devices [bluewavenet]
  * Add - increased wait_for_interface timeout from 10 to 120 seconds [bluewavenet]
  * Fix - don't do uci revert in case mesh11sd is installed [bluewavenet]
  * Fix - Remove escaped quotes in user agent field. [d1dee]
  * Fix - missing comma in awk command [bluewavenet]

 -- Rob White <dot@blue-wave.net>  Mon, 20 Apr 2026 14:15:07 +0000

opennds (10.3.1)
  * Fix - correctly process custom string in binauth [bluewavenet]
  * Fix - remove deprecated opkg test [bluewavenet]
  * Fix - block ipv6 routing if present [bluewavenet]
  * Fix - add missing service_data() to procd script [bluewavenet]
  * Fix - mhd version check [bluewavenet]

src/conf.h
/********************************************************************\
 * This program is free software; you can redistribute it and/or    *
 * modify it under the terms of the GNU General Public License as   *
 * published by the Free Software Foundation; either version 2 of   *
 * the License, or (at your option) any later version.              *
 *                                                                  *
 * This program is distributed in the hope that it will be useful,  *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of   *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the    *
 * GNU General Public License for more details.                     *
 *                                                                  *
 * You should have received a copy of the GNU General Public License*
 * along with this program; if not, contact:                        *
 *                                                                  *
 * Free Software Foundation           Voice:  +1-617-542-5942       *
 * 59 Temple Place - Suite 330        Fax:    +1-617-542-2652       *
 * Boston, MA  02111-1307,  USA       gnu@gnu.org                   *
 *                                                                  *
\********************************************************************/

/** @file conf.h
    @brief Config file parsing
    @author Copyright (C) 2004 Philippe April <papril777@yahoo.com>
    @author Copyright (C) 2007 Paul Kube <nodogsplash@kokoro.ucsd.edu>
    @author Copyright (C) 2015-2026 Modifications and additions by BlueWave Projects and Services <opennds@blue-wave.net>
*/

#define COPYRIGHT "openNDS, Copyright (C) 2015-2026 Modifications and additions by BlueWave Projects and Services"

#ifndef _CONF_H_
#define _CONF_H_

#define VERSION "11.0.0"

/*
 * Defines how many times should we try detecting the interface with the default route (in seconds).
 * If set to 0, it will keep retrying forever.
 */
#define NUM_EXT_INTERFACE_DETECT_RETRY 0

// How long we should wait per try to detect the interface with the default route if it isn't up yet (interval in seconds)
#define EXT_INTERFACE_DETECT_RETRY_INTERVAL 1

// Defaults configuration values
#ifndef SYSCONFDIR
#define DEFAULT_CONFIGFILE "/etc/opennds/opennds.conf"
#else
#define DEFAULT_CONFIGFILE SYSCONFDIR"/opennds/opennds.conf"
#endif
#define DEFAULT_DAEMON "0"
#define DEFAULT_ENABLED "1"
#define DEFAULT_DEBUGLEVEL "1"
#define DEFAULT_MAXCLIENTS "250"
#define DEFAULT_ONLINE_STATUS "0"
#define DEFAULT_GATEWAYINTERFACE "br-lan"
#define DEFAULT_GATEWAY_IPRANGE "0.0.0.0/0"
#define DEFAULT_GATEWAYNAME "openNDS"
#define DEFAULT_ENABLE_SERIAL_NUMBER_SUFFIX "1"
#define DEFAULT_GATEWAYPORT "2050"
#define DEFAULT_GATEWAYFQDN "status.client"
#define DEFAULT_DHCP_DEFAULT_URL_ENABLE "1"
#define DEFAULT_STATUSPATH "/usr/lib/opennds/client_params.sh"
#define DEFAULT_LOG_MOUNTPOINT "/tmp"
#define DEFAULT_MAX_PAGE_SIZE "10240"
#define DEFAULT_FASPORT "443"
#define DEFAULT_LOGIN_OPTION_ENABLED "1"
#define DEFAULT_MAX_LOG_ENTRIES "100"
#define DEFAULT_USE_OUTDATED_MHD "0"
#define DEFAULT_ALLOW_PREEMPTIVE_AUTHENTICATION "1"
#define DEFAULT_FAS_SECURE_ENABLED "1"
#define DEFAULT_FASPATH "/"
#define DEFAULT_FASKEY ""
#define DEFAULT_BINAUTH "/usr/lib/opennds/binauth_log.sh"
#define DEFAULT_CUSTOMBINAUTH "/usr/lib/opennds/custombinauth.sh"
#define DEFAULT_CHECKINTERVAL "15"
#define DEFAULT_SESSIONTIMEOUT "1440"
#define DEFAULT_PREAUTH_IDLE_TIMEOUT "30"
#define DEFAULT_AUTH_IDLE_TIMEOUT "120"
#define DEFAULT_REMOTES_REFRESH_INTERVAL "0"
#define DEFAULT_WEBROOT "/etc/opennds/htdocs"
#define DEFAULT_TMPFSMOUNTPOINT "/tmp"
#define DEFAULT_AUTHDIR "opennds_auth"
#define DEFAULT_DENYDIR "opennds_deny"
#define DEFAULT_PREAUTHDIR "opennds_preauth"
#define DEFAULT_SET_MSS "1" //allow setting the TCP Maximum Segment Size
#define DEFAULT_MSS_VALUE "0" // value to set the MSS. 0 means use max possible ie clamp-mss-to-pmtu
#define DEFAULT_RATE_CHECK_WINDOW "2" // The data rate check moving average window size multiply this by CHECKINTERVAL to give window size (or burst interval) in seconds
#define DEFAULT_UPLOAD_RATE "0" // 0 means no limit
#define DEFAULT_DOWNLOAD_RATE "0" // 0 means no limit
#define DEFAULT_UPLOAD_BUCKET_RATIO "1" // Allows control of upload rate limit threshold overrun per client
#define DEFAULT_DOWNLOAD_BUCKET_RATIO "1" // Allows control of download rate limit threshold overrun per client
#define DEFAULT_MAX_UPLOAD_BUCKET_SIZE "250" // Allows control over upload rate limiting packet loss at the expense of increased latency
#define DEFAULT_MAX_DOWNLOAD_BUCKET_SIZE "250" // Allows control over download rate limiting packet loss at the expense of increased latency
#define DEFAULT_UPLOAD_QUOTA "0" // 0 means no limit
#define DEFAULT_DOWNLOAD_QUOTA "0" // 0 means no limit
#define DEFAULT_FUP_UPLOAD_THROTTLE_RATE "0" // 0 means BLOCK the client until deauthed
#define DEFAULT_FUP_DOWNLOAD_THROTTLE_RATE "0" // 0 means BLOCK the client until deauthed
#define DEFAULT_UPLOAD_UNRESTRICTED_BURSTING "0" // 0 means disabled, 1 means enabled
#define DEFAULT_DOWNLOAD_UNRESTRICTED_BURSTING "0" // 0 means disabled, 1 means enabled
#define DEFAULT_NDSCTL_SOCK "ndsctl.sock"
#define DEFAULT_FW_MARK_AUTHENTICATED "0x30000"
#define DEFAULT_FW_MARK_AUTH_BLOCKED "0x30001"
#define DEFAULT_AUTHENTICATION_MARK "0x00030000"
#define DEFAULT_FW_MARK_TRUSTED "0x20000"
#define DEFAULT_THEMESPEC_PATH ""
#define DEFAULT_FAS_REMOTEFQDN "disabled"
#define DEFAULT_FAS_REMOTEIP "disabled"
#define DEFAULT_FAS_SSL "wget"

/* N.B.: default policies here must be ACCEPT, REJECT, or RETURN
 * In the .conf file, they must be allow, block, or passthrough
 * Mapping between these enforced by parse_empty_ruleset_policy()
 */
#define DEFAULT_EMPTY_TRUSTED_USERS_POLICY "accept"
#define DEFAULT_EMPTY_TRUSTED_USERS_TO_ROUTER_POLICY "accept"
#define DEFAULT_EMPTY_USERS_TO_ROUTER_POLICY "reject"
#define DEFAULT_EMPTY_AUTHENTICATED_USERS_POLICY "return"
#define DEFAULT_EMPTY_PREAUTHENTICATED_USERS_POLICY "reject"
#define DEFAULT_IP6 0

// Default lists
#define DEFAULT_TRUSTEDMACLIST ""
#define DEFAULT_FAS_CUSTOM_PARAMETERS_LIST ""
#define DEFAULT_FAS_CUSTOM_VARIABLES_LIST ""
#define DEFAULT_FAS_CUSTOM_IMAGES_LIST ""
#define DEFAULT_FAS_CUSTOM_FILES_LIST ""
#define DEFAULT_USERS_TO_ROUTER "allow%20udp%20port%2053 allow%20udp%20port%2067 allow%20tcp%20port%2022 allow%20tcp%20port%20443"
#define DEFAULT_AUTHENTICATED_USERS "allow%20all"
#define DEFAULT_PREAUTHENTICATED_USERS ""

// Firewall targets
typedef enum {
	TARGET_DROP,
	TARGET_REJECT,
	TARGET_ACCEPT,
	TARGET_RETURN,
	TARGET_LOG,
	TARGET_ULOG
} t_firewall_target;

// Firewall rules
typedef struct _firewall_rule_t {
	t_firewall_target target;	//@brief t_firewall_target
	char *protocol;		//@brief tcp, udp, etc ...
	char *port;			//@brief Port to block/allow
	char *mask;			//@brief Mask for the rule *destination*
	char *ipset;			//@brief IPset rule
	struct _firewall_rule_t *next;
} t_firewall_rule;

// MAC Addresses
typedef struct _MAC_t {
	char *mac;
	struct _MAC_t *next;
} t_MAC;

// Walled Garden Ports
typedef struct _WGP_t {
	unsigned int wgport;
	struct _WGP_t *next;
} t_WGP;

// Walled Garden FQDNs
typedef struct _WGFQDN_t {
	char *wgfqdn;
	struct _WGFQDN_t *next;
} t_WGFQDN;

// Custom FAS Parameters
typedef struct _FASPARAM_t {
	char *fasparam;
	struct _FASPARAM_t *next;
} t_FASPARAM;

// Custom FAS Variables
typedef struct _FASVAR_t {
	char *fasvar;
	struct _FASVAR_t *next;
} t_FASVAR;

// Custom FAS Images
typedef struct _FASIMG_t {
	char *fasimg;
	struct _FASIMG_t *next;
} t_FASIMG;

// Custom FAS Files
typedef struct _FASFILE_t {
	char *fasfile;
	struct _FASFILE_t *next;
} t_FASFILE;

// Configuration structure
typedef struct {
	char configfile[255];					//@brief name of the config file
	char *ndsctl_sock;					//@brief ndsctl path to socket
	char *internal_sock;					//@brief internal path to socket
	int enabled;						//@brief if openNDS is enabled
	int daemon;						//@brief if daemon != 0, use daemon mode
	int debuglevel;						//@brief Debug information verbosity
	int maxclients;						//@brief Maximum number of clients allowed
	int online_status;					//@brief Online status of the router, 1=online, 0=offline
	char *gw_name;						//@brief Name of the gateway; e.g. its SSID or a unique identifier for use in a remote FAS
	int enable_serial_number_suffix;			//@brief Enable/disable serial number suffix to gateway name
	char *http_encoded_gw_name;				//@brief http encoded name of the gateway, used as a templated variable in splash.htm
	char *url_encoded_gw_name;				//@brief url encoded name of the gateway used as variable in Preauth
	char *gw_interface;					//@brief Interface we will manage
	char *ext_gateway;					//@brief The interfaces and IP addresses of upstream gateways
	char *gw_iprange;					//@brief IP range on gw_interface we will manage
	char *gw_ip;						//@brief Internal IP (v4 or v6) for our web server
	char *gw_address;					//@brief Internal IP with port for our web server
	char *gw_mac;						//@brief MAC address of the interface we manage
	char *gw_fqdn;						//@brief FQDN of the client status page
	char *status_path;					//@brief Path to the client status page script
	int dhcp_default_url_enable;				//@brief Enable DHCP default-url (code 114 - RFC8910)
	unsigned int gw_port;					//@brief Port the webserver will run on
	unsigned int fas_port;					//@brief Port the fas server will run on
	int login_option_enabled;				//@brief Use default PreAuth Login script
	unsigned long long int max_log_entries;			//@brief set the maximum number of log entries
	int use_outdated_mhd;					//@brief Use outdated libmicrohttpd
	unsigned long long int max_page_size;			//@brief Max page size to be served by libmicrohttpd
	int allow_preemptive_authentication;			//@brief Allow Preemptive Authentication using the ndsctl utility
	int fas_secure_enabled;					//@brief Enable Secure FAS
	char *fas_path;						//@brief Path to forward authentication page of FAS
	char *fas_key;						//@brief AES key for FAS
	char *fas_remoteip;					//@brief IP addess of a remote FAS
	char *fas_remotefqdn;					//@brief FQDN of a remote FAS
	char *fas_url;						//@brief URL of a remote FAS
	char *fas_ssl;						//@brief SSL provider for FAS
	char *fas_hid;						//@brief Hash provider for FAS
	char *themespec_path;					//@brief Path to the ThemeSpec file to use for login_option_enabled = 3
	char *tmpfsmountpoint;					//@brief Mountpoint of the tmpfs drive eg /tmp etc.
	char *log_mountpoint;					//@brief Mountpoint of the log drive eg a USB drive mounted at /logs
	char *webroot;						//@brief Directory containing splash pages, etc.
	char *authdir;						//@brief Notional relative dir for authentication URL
	char *denydir;						//@brief Notional relative dir for denial URL
	char *preauthdir;					//@brief Notional relative dir for preauth URL
	int sessiontimeout;					//@brief Minutes of the default session length
	int preauth_idle_timeout;				//@brief Minutes a preauthenticated client will be kept in the system
	int auth_idle_timeout;					//@brief Minutes an authenticated client will be kept in the system
	int remotes_refresh_interval;				//@brief Minutes before remote files will be refreshed
	unsigned long long int remotes_last_refresh;		//@brief Time of last refresh of remote files
	int checkinterval;					//@brief Period the the client timeout check thread will run, in seconds
	int set_mss;						//@brief boolean, whether to set mss
	int mss_value;						//@brief int, mss value; <= 0 clamp to pmtu
	int rate_check_window;					//@brief window size in multiples of checkinterval for rate check moving average
	unsigned long long int download_rate;			//@brief Download rate, kb/s
	unsigned long long int upload_rate;			//@brief Upload rate, kb/s
	unsigned long long int download_bucket_ratio;		//@brief Allows control of download rate limit threshold overrun per client
	unsigned long long int upload_bucket_ratio;		//@brief Allows control of upload rate limit threshold overrun per client
	unsigned long long int max_upload_bucket_size;		//@brief control upload rate limiting packet loss at the expense of increased latency
	unsigned long long int max_download_bucket_size;	//@brief control download rate limiting packet loss at the expense of increased latency
	unsigned long long int download_quota;			//@brief Download quota, kB
	unsigned long long int upload_quota;			//@brief Upload quota, kB
	unsigned long long int fup_download_throttle_rate;	//@brief Fair Useage Policy Download throttle rate, kb/s, activated when quota exceeded
	unsigned long long int fup_upload_throttle_rate;	//@brief Fair Usage Policy Upload throttle rate, kb/s, activated when quota exceeded
	int download_unrestricted_bursting;			//@brief Enable/disable unrestriced bursting
	int upload_unrestricted_bursting;			//@brief Enable/disable unrestriced bursting
	int syslog_facility;					//@brief facility to use when using syslog for logging
	int macmechanism; 					//@brief mechanism wrt MAC addrs
	t_MAC *trustedmaclist;					//@brief list of trusted macs
	t_FASPARAM *fas_custom_parameters_list;			//@brief list of Custom FAS parameters
	t_FASVAR *fas_custom_variables_list;			//@brief list of Custom FAS variables
	t_FASIMG *fas_custom_images_list;			//@brief list of Custom FAS images
	t_FASFILE *fas_custom_files_list;			//@brief list of Custom FAS files
	char *custom_params;					//@brief FAS custom parameter string
	char *custom_vars;					//@brief FAS custom variable string
	char *custom_images;					//@brief FAS custom image string
	char *custom_files;					//@brief FAS custom file string
	unsigned int fw_mark_authenticated;			//@brief nftables mark for authenticated packets
	unsigned int fw_mark_auth_blocked;			//@brief nftables mark for auth_blocked packets
	char *authentication_mark;				//@brief Padded authentication mark
	unsigned int fw_mark_trusted;				//@brief nftables mark for trusted packets
	int ip6;						//@brief enable IPv6
	char *binauth;						//@brief external postauthentication program
	char *custombinauth;					//@brief external custom postauthentication program
	char *preauth;						//@brief external preauthentication program
	int lockfd;						//@brief ndsctl lockfile file descriptor
} s_config;

// @brief Get the current gateway configuration
s_config *config_get_config(void);

// @brief Initialise the conf system
void config_init(int argc, char **argv);
void parse_trusted_mac_list(const char[]);
void parse_fas_custom_parameters_list(const char[]);
void parse_fas_custom_variables_list(const char[]);
void parse_fas_custom_images_list(const char[]);
void parse_fas_custom_files_list(const char[]);

int is_trusted_mac(const char *mac);

int remove_from_trusted_mac_list(const char possiblemac[]);
int add_to_trusted_mac_list(const char possiblemac[]);

int check_ip_format(const char[]);
int check_mac_format(const char[]);

// config API, used in commandline.c
int set_debuglevel(const char[]);

#define LOCK_CONFIG() do { \
	debug(LOG_DEBUG, "Locking config"); \
	pthread_mutex_lock(&config_mutex); \
	debug(LOG_DEBUG, "Config locked"); \
} while (0)

#define UNLOCK_CONFIG() do { \
	debug(LOG_DEBUG, "Unlocking config"); \
	pthread_mutex_unlock(&config_mutex); \
	debug(LOG_DEBUG, "Config unlocked"); \
} while (0)

#endif // _CONF_H_

char *set_list_str(char *list, const char *default_list, char *debug_level);
